include/roaring/memory.h

```diff
@@ -0,0 +1,39 @@
+#ifndef INCLUDE_ROARING_MEMORY_H_
+#define INCLUDE_ROARING_MEMORY_H_
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#include <stddef.h>  // for size_t
+
+typedef void* (*p_roaring_malloc)(size_t);
+typedef void* (*p_roaring_realloc)(void*, size_t);
+typedef void* (*p_roaring_calloc)(size_t, size_t);
+typedef void (*p_roaring_free)(void*);
+typedef void* (*p_roaring_aligned_malloc)(size_t, size_t);
+typedef void (*p_roaring_aligned_free)(void*);
+
+typedef struct roaring_memory_s {
+    p_roaring_malloc malloc;
+    p_roaring_realloc realloc;
+    p_roaring_calloc calloc;
+    p_roaring_free free;
+    p_roaring_aligned_malloc aligned_malloc;
+    p_roaring_aligned_free aligned_free;
+} roaring_memory_t;
+
+void init_roaring_memory_hook(roaring_memory_t memory_hook);
+
+void* roaring_malloc(size_t);
+void* roaring_realloc(void*, size_t);
+void* roaring_calloc(size_t, size_t);
+void roaring_free(void*);
+void* roaring_aligned_malloc(size_t, size_t);
+void roaring_aligned_free(void*);
+
+#ifdef __cplusplus
+}
+#endif
+
```

Comment:
Since we require C11/C++11, we could `#define` the standard allocator functions to a failing static assert or something similar. This makes it easier to catch explicit uses in our implementation. We would `#undef` them in `memory.c` only to allow its use there.
I'm not sure this is doable tho, it may end up in public headers and break users' code.

Reply:
The idea makes sense to me that it would avoid some 'wild' malloc function. 
But yes, if some users just want to use simple malloc-free functions or include this lib in another project, this would definitely break their code.

Reply:
So IMHO, I would choose not to do this to avoid influence to other code.

Reply:
Agreed.